// @near-eth/nep141-erc20/dist/natural-erc20/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.findAllTransfers = exports.findAllTransactions = exports.recover = exports.checkApprove = exports.approve = exports.sendToNear = exports.getBalance = exports.getAllowance = exports.getMetadata = void 0;
var getMetadata_1 = require("./getMetadata");
Object.defineProperty(exports, "getMetadata", { enumerable: true, get: function () { return __importDefault(getMetadata_1).default; } });
var getAllowance_1 = require("./getAllowance");
Object.defineProperty(exports, "getAllowance", { enumerable: true, get: function () { return __importDefault(getAllowance_1).default; } });
var getBalance_1 = require("./getBalance");
Object.defineProperty(exports, "getBalance", { enumerable: true, get: function () { return __importDefault(getBalance_1).default; } });
var sendToNear_1 = require("./sendToNear");
Object.defineProperty(exports, "sendToNear", { enumerable: true, get: function () { return sendToNear_1.initiate; } });
Object.defineProperty(exports, "approve", { enumerable: true, get: function () { return sendToNear_1.approve; } });
Object.defineProperty(exports, "checkApprove", { enumerable: true, get: function () { return sendToNear_1.checkApprove; } });
Object.defineProperty(exports, "recover", { enumerable: true, get: function () { return sendToNear_1.recover; } });
Object.defineProperty(exports, "findAllTransactions", { enumerable: true, get: function () { return sendToNear_1.findAllTransactions; } });
Object.defineProperty(exports, "findAllTransfers", { enumerable: true, get: function () { return sendToNear_1.findAllTransfers; } });
